### DISCUSSION TITLE
SCM Extension Questions on nesting groups, identifying the actual resource state, and refreshing the scm view

### DISCUSSION THREAD

**Discussion by GallifreyMage** [microsoft/vscode-discussions#2427] [Extension Development QnA]
Hi there,
I've posted once previously.  In the mean time I've used AI to get me pretty far into building my SCM extension.
I can parse my existing repo, build resourceStates and resourceGroups, and have them show up in my SCM tree view.
However, I am struggling with several things, but here's the lay of the land:

Our repositories often have >1000 source files.  Most of them are not changed ant any given time.  In fact in our case we have to extract them from the repository to the local workspace in order to edit them.  But seeing all the files in the SCM tree is extremely useful, even if I'm not changing them.

1. How do I distinguish that a given resource is changed or unchanged?

2. When I extract a resource to my local workspace, how do I reference both the local edit and the original so I can diff them?  Can the resourceState accommodate this?

3. When I change or dispose of resourceStates, it seems the SCM view never updates.  Is this a known issue?  I can however add stuff as much as I want and it just gets appended to the end of the list.  It's like something in the sourceControl object isn't getting reset or the SCM view itself doesn't clear itself out appropriately.

4. Because we have so many files, I'm using the resourceGroups to act as a way to filter or organize the list.  I can construct them based on the folder structure in the repository, or based on other groupings (like tags).  But, when I create a new set of groups to change the organization of the list, I am not able to clear out the previous set of groups.  I have called dispose() on them but they still show up in the SCM view.  I've even tried going through each group and clearing out the resourceStates array, but that doesn't change anything either.

5. Due to the sheer volume of groups, and files, it would be useful to be able to nest groups (a la a nested folder structure).  Is there an easy way to do this? 

6. I have a lot of other meta-data available to me for each resource (file) - is there a way I can extend the resourceState object to add the additional info to make it easier to use?

I think that's a good start.  Any help on any of the above subjects would be most welcome.

Many thanks,
GallifreyMage

**Comment by robconery** [comment]
Are you building your own SCM (assuming you mean Source Control Management) system? If so: are you wanting to hook into the existing Source Control menu? If you could add some context it would be helpful.
